ENH: Implement classifier orthogonality in FIC plot

tylerjereddy opened the pull request:
This is the GitHub version of the in-house MR from @adamwitmer at https://lisdi-git.lanl.gov/treddy/ldrd_virus_work/-/merge_requests/88. More detailed information and discussion is available there. I've merged the GitHub version of `awitmer_fix_cv_agg` into this branch and resolved merge conflicts manually.

From that in-house MR, the description of this PR is copied below:

* Implements FIC plot design that includes counts of important features from both classifier and SHAP importances
* Proposes method of displaying `response effect` symbol to right of FIC bar along with percent surface exposed
* Changes data structure handling to allow for addition of multiple classifiers
* fixes a bug in train/test split by setting `StratifiedKFold` `shuffle=True` and setting `random_state` to ensure that the same splits are used for all classifiers

To run the workflow:

1. cd to a new dir called `[target_column]`
2. run the workflow using the following command:

```
python ../viral_seq/run_workflow.py --cache extract --cache-tarball dtra_cache.tar.gz --optimize none --train-file receptor_training.csv --test-file none --workflow DTRA --target-column [target_column] --mapping-method jurgen_schmidt
```

A sample of the new stacked FIC plot produced is shown below the fold for the integrin target:

<details>

<img width="2400" height="3000" alt="image" src="https://github.com/user-attachments/assets/d5390a48-2614-48f1-a9cd-486e21232967" />

</details>

TODO list is probably:

- [x] check for any problems due to merge conflict resolution
- [x] make sure CI is passing on GitHub
- [x] run a sample DTRA workflow locally and check the output plots
- [x] a problem with an extraneous space was mentioned at https://lisdi-git.lanl.gov/treddy/ldrd_virus_work/-/merge_requests/88#note_35199
- [x] make sure there is a sensible resolution to discussion at https://lisdi-git.lanl.gov/treddy/ldrd_virus_work/-/merge_requests/88#note_106024 about `test_pearson_aggregation`
- [x] double check test line coverage

@austinschneider thought the in-house version of this PR was in pretty good shape, but let's be cautious here with the repo change and branches in flux...

ajwitmer committed:
ENH: Implement classifier orthogonality in FIC plot

MAINT: Fix plots to display percentages

MAINT: Integrate orthogonality with FIC plot after rebase

- add 'plt.close()' to end of 'plot_roc' in 'spillover_predict.py' that was causing an error in pytest plot
- modify tests to account for new 'orthogonal' data structure

MAINT: fix conflicts after rebase

MAINT, ENH, TST: Address reviewer comments

- Reformatted diff to fix erroneous code changes
- Fixed docs to reflect new changes
- Fixed mypy type hints
- Moved data munging operations to appropriate functions
- Abstracted feature dataframe sorting to function with test
- Fixed FIC plot titles/legends for clarity
- Modified FIC Plots to display color bars
- Added FIC plot test case to display estimator disagreement

MAINT: fix conflicts after rebase

- modify functions/tests to integrate feature counting with downstream data handling
- improve logic for feature ranking by adding secondary rank against Pearson R values

MAINT: address reviewer comments

- improve plotting logic for percent values
- parametrize values in test
- modify input argument for feature_signs
- remove unrelated changes
- fix docs and improve readability of code

MAINT: modify workflow to handle multiple classifiers

MAINT: address reviewer comments

- modify FIC legend location handling
- add test case for multiple classifiers

MAINT: modify functions/tests to account for multiple classifiers

MAINT: address reviewer comments

- modify test for plotting ROC of multiple classifiers
- fix foramtting error
- change wording of TODO line comment

MAINT: fix label syntax

MAINT: address reviewer comments

- fix docs for clarity
- move functions/tests to separate library modules
- improve test logic and add assertions for test coverage
- fix unrelated whitespace changes
- fix mypy type hints
- fix typos

MAINT: modify logic for ranking features to weight positive SHAP responses

- fix tests associated with change in ranking

MAINT: address reviewer comments and fix conflicts after rebase

- fix tests
- update docs for clarity
- remove unnecessary variable assignments and array coercions
- remove duplicated function definitions

MAINT: fix conflicts after rebase

MAINT: revert unnecessary changes related to order of ``feature_count`` data structure

- add CLI option for ``--max_features``

STY: fix unrelated whitespace changes

MAINT, BUG: address reviewer comments

- change `proportion` to `percentage`
- add doc to test for clarity
- improve conciseness of list generation
- raise ValueError in FIC_plot and add test
- remove duplicate rng initialization
- fix bug; not shuffling dataset
- update expected values to accommodate bug fix

MAINT: fix conflicts after rebase

ajwitmer committed:
TST: improve assertions for properties of `test_pearson_aggregation`

tylerjereddy committed:
Merge remote-tracking branch 'github/awitmer_fix_cv_agg' into awitmer_clfr_orthog

tylerjereddy committed:
MAINT, BUG: fixups after merge

austinschneider commented:
Thanks for porting this over @tylerjereddy, I'll give it a careful look through.

tylerjereddy commented:
Now there are some merge conflicts here--let me see if I can resolve those in a way that at least passes the testsuite.

I believe the suggestion was that after this PR gets merged, we can try to merge `awitmer_fix_cv_agg` into the DTRA development branch (`amallela_test_ml_workflow_2`), which would make things a lot easier moving forward.

tylerjereddy committed:
Merge branch 'awitmer_fix_cv_agg' into awitmer_clfr_orthog

tylerjereddy committed:
MAINT: PR 7 revisions

* Rough attempts at cleanup after merge conflict
resolution.

tylerjereddy commented on viral_seq/tests/test_workflow.py:
```diff
@@ -1101,12 +1146,31 @@ def test_pearson_aggregation():
-    ]
-    # check first four pearson values, numbers have tendency to vary slightly based on dependency versions
-    assert_allclose(pearson_rank[: len(pearson_rank_exp)], pearson_rank_exp, rtol=6e-6)
+    # expected property of the test is ranking kmer_0 > kmer_1 > kmer_3
```
this doesn't mesh with the discussion at https://lisdi-git.lanl.gov/treddy/ldrd_virus_work/-/merge_requests/88#note_106024, which indicates:

> it should remain enforceable that this order is always true in the top 4: [`kmer_0` > `kmer_2` > `kmer_3`]

> `kmer_1` is very close in absolute value of correlation to `kmer_0`, so depending on exact shuffling I can see why its magnitude in the target branch actually exceeds `kmer_0`.

In any case, looks like 2 and 3 swapping sometimes when I change seeds, so maybe there is a bug to work out.

tylerjereddy committed:
MAINT: PR 7 revisions

* Fix a grammar issue in `test_pearson_aggregation()`, based
on reviewer feedback.

* Adjustments to `test_pearson_aggregation()` to get it to pass
(though I'm not really happy with it yet...).

tylerjereddy commented:
I pushed up a revision that allows `test_pearson_aggregation()` to pass locally. I'm not super happy with it, but I believe it does increase the robustness a bit. The guarantees are a bit weaker now, but the general ranking/trend appears to be correct--that may be as well as I can do without postdoc support to drive it forward. I'll keep an eye on that test...

There are still two more test failures to address here.

tylerjereddy committed:
TST: PR 7 revisions

* Fix an incorrectly sized `feature_rank_array` in `test_train_clfr()`.

tylerjereddy committed:
TST: PR 7 revisions

* `test_train_clfr()` cleanups--the new test results are either
better or acceptable within random variation; added comments
to support them

tylerjereddy commented:
I pushed in some fixes for `test_train_clfr()` -- these I was much more comfortable with -- the changes were either straight up improvements in the test results backed up by inline comments, or test result mutations that are acceptable within random noise.

tylerjereddy committed:
MAINT: PR 7 revisions

* Remove an extraneous space mentioned during the code review
at: https://lisdi-git.lanl.gov/treddy/ldrd_virus_work/-/merge_requests/88#note_35199

tylerjereddy commented on viral_seq/run_workflow.py:
```diff
@@ -636,135 +652,133 @@ def FIC_plot(
+    )
+    # TODO: how to calculate `n_classifiers` here?
+    # leftover from merge conflict resolution...
+    n_classifiers = 1
```
Still need to deal with this, I think.

tylerjereddy commented on viral_seq/run_workflow.py:
```diff
@@ -636,135 +652,133 @@ def FIC_plot(
+    ax.set_xlabel(f"% ML models where ranked in top {max_features} features")
+    ax.set_title(
+        f"Feature importance consensus amongst\n {n_folds} folds for {target_name} binding"
+    )
```
Actually, isn't this `set_title` overriding the one I'm talking about above at https://github.com/lanl/ldrd_virus_work/pull/7/files#r2353864590? Confusion because of merge conflict resolution?

tylerjereddy commented on viral_seq/tests/test_workflow.py:
Today, I've re-reviewed all files except for this one (and didn't see any major new/obvious problems)--try to wrap that review up this evening, then check test line coverage and do a workflow run that exercises both the new stacked plot approach + multiple random seeds (`--n_seeds` for DTRA workflow I think)

tylerjereddy commented on viral_seq/tests/test_workflow.py:
```diff
@@ -1101,12 +1161,35 @@ def test_pearson_aggregation():
+    feature_df["SHAP_1"] = [1, 0, 4]
+    feature_df["Pearson R"] = [0.99, -0.99, -0.41]
+
+    out_df = feature_importance.sort_feature_counts(feature_df, n_folds=5)
```
It seems fairly clear to me that this test should reside in `test_feature_importance.py`, not sure why it was placed here.

tylerjereddy commented on viral_seq/tests/test_workflow.py:
```diff
@@ -1101,12 +1161,35 @@ def test_pearson_aggregation():
+    out_df = feature_importance.sort_feature_counts(feature_df, n_folds=5)
+    out_features = np.asarray(out_df["Features"])
+    out_exp = kmer_features[np.asarray([2, 0, 1])]
+    assert_array_equal(out_exp, out_features)
```
The order of arguments here is reversed--I mentioned this several times over the past year--the documented function signature is:

`assert_array_equal(actual, desired)`

I'll fix it myself now...

tylerjereddy committed:
TST, MAINT: PR 7 revisions

* Migrate `test_sort_feature_counts()` to the appropriate
testing module, and fix the assertion order convention.

* Fix up the FIC plot titles to include both the number
of seeds and the number of folds (separately).

tylerjereddy commented:
I think the only remaining item is to run the workflow locally on this branch (+ new random seed feature). Trying that now with this incantation: `time python ../viral_seq/run_workflow.py --cache extract --cache-tarball dtra_cache.tar.gz --optimize none --train-file receptor_training.csv --test-file none --workflow DTRA --target-column IN --mapping-method jurgen_schmidt --n_seeds 3`

tylerjereddy commented:
The above workflow incantation completed on my ARM Mac laptop in `41:56.26`, with successful exit code `0`.  However, there are problems to be addressed before we can merge here.

The FIC plot, which is arguably our most important summary results plot, has a multitude of display issues, including surface exposed %s running off the side of the plot, y-axis kmer labels being cut off, and the legend being cut off. See sample result from above incantation below the fold.

<details>

<img width="2400" height="3000" alt="FIC_Integrin" src="https://github.com/user-attachments/assets/a76e37a5-a57d-4cea-9bd3-dfac694ca44a" />


</details>

tylerjereddy committed:
BUG: PR 7 revisions

* Add a regression test and bug fix for an issue where
`train_clfr()` was not correctly normalizing feature
count percentages to account for the number of seeds used.

tylerjereddy commented:
I pushed in a regression test and tentative bug fix for `train_clfr()` failing to properly account for the number of seeds used when normalizing the % occurrences of the kmer features.

I'll probably have to rerun the above workflow incantation from scratch now, but let me check if I can do a faster run to check a bit more quickly initially.

tylerjereddy commented:
Ok, I got this branch setup/tested in a `gp160` `screen` session and I've started the workflow incantation from https://github.com/lanl/ldrd_virus_work/pull/7#issuecomment-3301246079 from scratch there.

Fingers crossed the FIC plot looks more sensible after the latest revisions here.

tylerjereddy commented:
Ok, with my bug fix pushed in, the workflow incantation above produced a normal-looking FIC plot after running for 65 minutes on `gp160`:

<img width="2400" height="3000" alt="image" src="https://github.com/user-attachments/assets/f98cf081-f0ac-437c-a788-3b4c15157151" />

That's as good as I can do here without postdoctoral support, in it goes.
